### ISSUE TITLE
[React Native] undefined languages warning when using iOS Simulator

### ISSUE THREAD

**Issue by azuken** [i18next/react-i18next#1303]
## 🐛 Bug Report

When initialising i18next on React Native app on iOS simulator, warning appears (translations and display is working fine). On Android emulator / real device, no warning is appearing. The "quick fix" is to set `lng` config option, but with this option detector is not working anymore and force language.

The warning :

`WARN  react-i18next:: i18n.languages were undefined or empty undefined`

With debug set to true, I have another warning which says this : 

` WARN  i18next::translator: key "screenTitles.trackingOrderTitle" for languages "en" won't get resolved as namespace "translation" was not yet loaded This means something IS WRONG in your setup. You access the t function before i18next.init / i18next.loadNamespace / i18next.changeLanguage was done. Wait for the callback or Promise to resolve before accessing it!!! 
`

I've set this condition on my App root to be sure that i18n is initialised before using any translation hook : 

```js
const AppConfig = () => {
  return i18n.isInitialized ? (
      <AppContainer />
  ) : (
    <></>
  )
}

```

I've also tried to use init callback to set a boolean state, same behaviour.

## To Reproduce

```js
  import i18n from 'i18next'
  import i18nextReactNative from 'i18next-react-native-language-detector'
  import enTranslation from './en/translation.json'
  import frTranslation from './fr/translation.json'
  import esTranslation from './es/translation.json'
  import deTranslation from './de/translation.json'
  import { initReactI18next } from 'react-i18next'

  const resources = {
    en: { translation: enTranslation },
    fr: { translation: frTranslation },
    es: { translation: esTranslation },
    de: { translation: deTranslation },
  }

  void i18n
    .use(initReactI18next)
    .use(i18nextReactNative)
    .init({
      resources,
      fallbackLng: 'en',
      interpolation: {
        escapeValue: false,
      },
    })

export default i18n
```

## Expected behavior

No warning appearing

## Your Environment

- React Native 0.64.0, React 17.0.2
- i18next 20.2.1 / react-i18next 11.8.13 / i18next-react-native-language-detector 1.0.2
- MacOS
- iOS Simulator "iPhone 12 Pro" with iOS 14.4

**Comment by adrai** [comment]
You may need a language detector: https://github.com/DylanVann/i18next-react-native-language-detector

Maybe compare also with this example: https://github.com/i18next/react-i18next/blob/master/example/react-native/App.js

**Comment by azuken** [comment]
@adrai As you can see in my sample code, I already have a language detector.

**Comment by adrai** [comment]
Sorry, dis not see it on my mobile...

Can you paste the complete i18next debug output?

**Comment by azuken** [comment]
Here is all log content I have when debug is enabled for i18next.

```
 WARN  react-i18next:: i18n.languages were undefined or empty undefined 
PreLoginScreen@http://localhost:8081/index.bundle?platform=ios&dev=true&minify=false&modulesOnly=false&runModule=true&app=com.xxx.xxx:316937:26
StaticContainer@http://localhost:8081/index.bundle?platform=ios&dev=true&minify=false&modulesOnly=false&runModule=true&app=com.xxx.xxx:139203:17
EnsureSingleNavigator@http://localhost:8081/index.bundle?platform=ios&dev=true&minify=false&modulesOnly=false&runModule=true&app=com.xxx.xxx:137851:24
SceneView@http://localhost:8081/index.bundle?platform=ios&dev=true&minify=false&modulesOnly=false&runModule=true&app=com.xxx.xxx:139093:22
RCTView
View
RCTView
View
RCTView
View
CardSheet@http://localhost:8081/index.bundle?platform=ios&dev=true&minify=false&modulesOnly=false&runModule=true&app=com.xxx.xxx:316167:23
RCTView
View
AnimatedComponent@http://localhost:8081/index.bundle?platform=ios&dev=true&minify=false&modulesOnly=false&runModule=true&app=com.xxx.xxx:62612:24
AnimatedComponentWrapper
PanGestureHandler@http://localhost:8081/index.bundle?platform=ios&dev=true&minify=false&modulesOnly=false&runModule=true&app=com.xxx.xxx:309006:38
PanGestureHandler@http://localhost:8081/index.bundle?platform=ios&dev=true&minify=false&modulesOnly=false&runModule=true&app=com.xxx.xxx:308271:34
RCTView
View
AnimatedComponent@http://localhost:8081/index.bundle?platform=ios&dev=true&minify=false&modulesOnly=false&runModule=true&app=com.xxx.xxx:62612:24
AnimatedComponentWrapper
RCTView
View
Card@http://localhost:8081/index.bundle?platform=ios&dev=true&minify=false&modulesOnly=false&runModule=true&app=com.xxx.xxx:313751:36
CardContainer@http://localhost:8081/index.bundle?platform=ios&dev=true&minify=false&modulesOnly=false&runModule=true&app=com.xxx.xxx:313428:22
RNSScreen
AnimatedComponent@http://localhost:8081/index.bundle?platform=ios&dev=true&minify=false&modulesOnly=false&runModule=true&app=com.xxx.xxx:62612:24
AnimatedComponentWrapper
Screen@http://localhost:8081/index.bundle?platform=ios&dev=true&minify=false&modulesOnly=false&runModule=true&app=com.xxx.xxx:311981:36
MaybeScreen@http://localhost:8081/index.bundle?platform=ios&dev=true&minify=false&modulesOnly=false&runModule=true&app=com.xxx.xxx:312330:24
RNSScreenContainer
ScreenContainer@http://localhost:8081/index.bundle?platform=ios&dev=true&minify=false&modulesOnly=false&runModule=true&app=com.xxx.xxx:312072:36
MaybeScreenContainer@http://localhost:8081/index.bundle?platform=ios&dev=true&minify=false&modulesOnly=false&runModule=true&app=com.xxx.xxx:312299:23
CardStack@http://localhost:8081/index.bundle?platform=ios&dev=true&minify=false&modulesOnly=false&runModule=true&app=com.xxx.xxx:311455:36
KeyboardManager@http://localhost:8081/index.bundle?platform=ios&dev=true&minify=false&modulesOnly=false&runModule=true&app=com.xxx.xxx:316336:36
RNCSafeAreaProvider
SafeAreaProvider@http://localhost:8081/index.bundle?platform=ios&dev=true&minify=false&modulesOnly=false&runModule=true&app=com.xxx.xxx:308043:24
SafeAreaProviderCompat@http://localhost:8081/index.bundle?platform=ios&dev=true&minify=false&modulesOnly=false&runModule=true&app=com.xxx.xxx:316778:24
RCTView
View
StackView@http://localhost:8081/index.bundle?platform=ios&dev=true&minify=false&modulesOnly=false&runModule=true&app=com.xxx.xxx:307507:36
StackNavigator@http://localhost:8081/index.bundle?platform=ios&dev=true&minify=false&modulesOnly=false&runModule=true&app=com.xxx.xxx:307388:32
AuthenticationStack
EnsureSingleNavigator@http://localhost:8081/index.bundle?platform=ios&dev=true&minify=false&modulesOnly=false&runModule=true&app=com.xxx.xxx:137851:24
BaseNavigationContainer@http://localhost:8081/index.bundle?platform=ios&dev=true&minify=false&modulesOnly=false&runModule=true&app=com.xxx.xxx:137485:28
ThemeProvider@http://localhost:8081/index.bundle?platform=ios&dev=true&minify=false&modulesOnly=false&runModule=true&app=com.xxx.xxx:141746:21
NavigationContainer@http://localhost:8081/index.bundle?platform=ios&dev=true&minify=false&modulesOnly=false&runModule=true&app=com.xxx.xxx:141652:26
Provider@http://localhost:8081/index.bundle?platform=ios&dev=true&minify=false&modulesOnly=false&runModule=true&app=com.xxx.xxx:109664:26
Provider@http://localhost:8081/index.bundle?platform=ios&dev=true&minify=false&modulesOnly=false&runModule=true&app=com.xxx.xxx:302526:26
StripeProvider@http://localhost:8081/index.bundle?platform=ios&dev=true&minify=false&modulesOnly=false&runModule=true&app=com.xxx.xxx:299862:41
App@http://localhost:8081/index.bundle?platform=ios&dev=true&minify=false&modulesOnly=false&runModule=true&app=com.xxx.xxx:149198:20
Provider@http://localhost:8081/index.bundle?platform=ios&dev=true&minify=false&modulesOnly=false&runModule=true&app=com.xxx.xxx:144062:26
AuthProvider@http://localhost:8081/index.bundle?platform=ios&dev=true&minify=false&modulesOnly=false&runModule=true&app=com.xxx.xxx:142753:65
AppContainer@http://localhost:8081/index.bundle?platform=ios&dev=true&minify=false&modulesOnly=false&runModule=true&app=com.xxx.xxx:142655:68
Provider@http://localhost:8081/index.bundle?platform=ios&dev=true&minify=false&modulesOnly=false&runModule=true&app=com.xxx.xxx:109664:26
AppConfig
RCTView
View
RCTView
View
AppContainer@http://localhost:8081/index.bundle?platform=ios&dev=true&minify=false&modulesOnly=false&runModule=true&app=com.xxx.xxx:69675:36
 WARN  i18next: hasLoadedNamespace: i18n.languages were undefined or empty undefined 
PreLoginScreen@http://localhost:8081/index.bundle?platform=ios&dev=true&minify=false&modulesOnly=false&runModule=true&app=com.xxx.xxx:316937:26
StaticContainer@http://localhost:8081/index.bundle?platform=ios&dev=true&minify=false&modulesOnly=false&runModule=true&app=com.xxx.xxx:139203:17
EnsureSingleNavigator@http://localhost:8081/index.bundle?platform=ios&dev=true&minify=false&modulesOnly=false&runModule=true&app=com.xxx.xxx:137851:24
SceneView@http://localhost:8081/index.bundle?platform=ios&dev=true&minify=false&modulesOnly=false&runModule=true&app=com.xxx.xxx:139093:22
RCTView
View
RCTView
View
RCTView
View
CardSheet@http://localhost:8081/index.bundle?platform=ios&dev=true&minify=false&modulesOnly=false&runModule=true&app=com.xxx.xxx:316167:23
RCTView
View
AnimatedComponent@http://localhost:8081/index.bundle?platform=ios&dev=true&minify=false&modulesOnly=false&runModule=true&app=com.xxx.xxx:62612:24
AnimatedComponentWrapper
PanGestureHandler@http://localhost:8081/index.bundle?platform=ios&dev=true&minify=false&modulesOnly=false&runModule=true&app=com.xxx.xxx:309006:38
PanGestureHandler@http://localhost:8081/index.bundle?platform=ios&dev=true&minify=false&modulesOnly=false&runModule=true&app=com.xxx.xxx:308271:34
RCTView
View
AnimatedComponent@http://localhost:8081/index.bundle?platform=ios&dev=true&minify=false&modulesOnly=false&runModule=true&app=com.xxx.xxx:62612:24
AnimatedComponentWrapper
RCTView
View
Card@http://localhost:8081/index.bundle?platform=ios&dev=true&minify=false&modulesOnly=false&runModule=true&app=com.xxx.xxx:313751:36
CardContainer@http://localhost:8081/index.bundle?platform=ios&dev=true&minify=false&modulesOnly=false&runModule=true&app=com.xxx.xxx:313428:22
RNSScreen
AnimatedComponent@http://localhost:8081/index.bundle?platform=ios&dev=true&minify=false&modulesOnly=false&runModule=true&app=com.xxx.xxx:62612:24
AnimatedComponentWrapper
Screen@http://localhost:8081/index.bundle?platform=ios&dev=true&minify=false&modulesOnly=false&runModule=true&app=com.xxx.xxx:311981:36
MaybeScreen@http://localhost:8081/index.bundle?platform=ios&dev=true&minify=false&modulesOnly=false&runModule=true&app=com.xxx.xxx:312330:24
RNSScreenContainer
ScreenContainer@http://localhost:8081/index.bundle?platform=ios&dev=true&minify=false&modulesOnly=false&runModule=true&app=com.xxx.xxx:312072:36
MaybeScreenContainer@http://localhost:8081/index.bundle?platform=ios&dev=true&minify=false&modulesOnly=false&runModule=true&app=com.xxx.xxx:312299:23
CardStack@http://localhost:8081/index.bundle?platform=ios&dev=true&minify=false&modulesOnly=false&runModule=true&app=com.xxx.xxx:311455:36
KeyboardManager@http://localhost:8081/index.bundle?platform=ios&dev=true&minify=false&modulesOnly=false&runModule=true&app=com.xxx.xxx:316336:36
RNCSafeAreaProvider
SafeAreaProvider@http://localhost:8081/index.bundle?platform=ios&dev=true&minify=false&modulesOnly=false&runModule=true&app=com.xxx.xxx:308043:24
SafeAreaProviderCompat@http://localhost:8081/index.bundle?platform=ios&dev=true&minify=false&modulesOnly=false&runModule=true&app=com.xxx.xxx:316778:24
RCTView
View
StackView@http://localhost:8081/index.bundle?platform=ios&dev=true&minify=false&modulesOnly=false&runModule=true&app=com.xxx.xxx:307507:36
StackNavigator@http://localhost:8081/index.bundle?platform=ios&dev=true&minify=false&modulesOnly=false&runModule=true&app=com.xxx.xxx:307388:32
AuthenticationStack
EnsureSingleNavigator@http://localhost:8081/index.bundle?platform=ios&dev=true&minify=false&modulesOnly=false&runModule=true&app=com.xxx.xxx:137851:24
BaseNavigationContainer@http://localhost:8081/index.bundle?platform=ios&dev=true&minify=false&modulesOnly=false&runModule=true&app=com.xxx.xxx:137485:28
ThemeProvider@http://localhost:8081/index.bundle?platform=ios&dev=true&minify=false&modulesOnly=false&runModule=true&app=com.xxx.xxx:141746:21
NavigationContainer@http://localhost:8081/index.bundle?platform=ios&dev=true&minify=false&modulesOnly=false&runModule=true&app=com.xxx.xxx:141652:26
Provider@http://localhost:8081/index.bundle?platform=ios&dev=true&minify=false&modulesOnly=false&runModule=true&app=com.xxx.xxx:109664:26
Provider@http://localhost:8081/index.bundle?platform=ios&dev=true&minify=false&modulesOnly=false&runModule=true&app=com.xxx.xxx:302526:26
StripeProvider@http://localhost:8081/index.bundle?platform=ios&dev=true&minify=false&modulesOnly=false&runModule=true&app=com.xxx.xxx:299862:41
App@http://localhost:8081/index.bundle?platform=ios&dev=true&minify=false&modulesOnly=false&runModule=true&app=com.xxx.xxx:149198:20
Provider@http://localhost:8081/index.bundle?platform=ios&dev=true&minify=false&modulesOnly=false&runModule=true&app=com.xxx.xxx:144062:26
AuthProvider@http://localhost:8081/index.bundle?platform=ios&dev=true&minify=false&modulesOnly=false&runModule=true&app=com.xxx.xxx:142753:65
AppContainer@http://localhost:8081/index.bundle?platform=ios&dev=true&minify=false&modulesOnly=false&runModule=true&app=com.xxx.xxx:142655:68
Provider@http://localhost:8081/index.bundle?platform=ios&dev=true&minify=false&modulesOnly=false&runModule=true&app=com.xxx.xxx:109664:26
AppConfig
RCTView
View
RCTView
View
AppContainer@http://localhost:8081/index.bundle?platform=ios&dev=true&minify=false&modulesOnly=false&runModule=true&app=com.xxx.xxx:69675:36
 WARN  i18next::translator: key "preLogin.signIn" for languages "en" won't get resolved as namespace "translation" was not yet loaded This means something IS WRONG in your setup. You access the t function before i18next.init / i18next.loadNamespace / i18next.changeLanguage was done. Wait for the callback or Promise to resolve before accessing it!!! 
PreLoginScreen@http://localhost:8081/index.bundle?platform=ios&dev=true&minify=false&modulesOnly=false&runModule=true&app=com.xxx.xxx:316937:26
StaticContainer@http://localhost:8081/index.bundle?platform=ios&dev=true&minify=false&modulesOnly=false&runModule=true&app=com.xxx.xxx:139203:17
EnsureSingleNavigator@http://localhost:8081/index.bundle?platform=ios&dev=true&minify=false&modulesOnly=false&runModule=true&app=com.xxx.xxx:137851:24
SceneView@http://localhost:8081/index.bundle?platform=ios&dev=true&minify=false&modulesOnly=false&runModule=true&app=com.xxx.xxx:139093:22
RCTView
View
RCTView
View
RCTView
View
CardSheet@http://localhost:8081/index.bundle?platform=ios&dev=true&minify=false&modulesOnly=false&runModule=true&app=com.xxx.xxx:316167:23
RCTView
View
AnimatedComponent@http://localhost:8081/index.bundle?platform=ios&dev=true&minify=false&modulesOnly=false&runModule=true&app=com.xxx.xxx:62612:24
AnimatedComponentWrapper
PanGestureHandler@http://localhost:8081/index.bundle?platform=ios&dev=true&minify=false&modulesOnly=false&runModule=true&app=com.xxx.xxx:309006:38
PanGestureHandler@http://localhost:8081/index.bundle?platform=ios&dev=true&minify=false&modulesOnly=false&runModule=true&app=com.xxx.xxx:308271:34
RCTView
View
AnimatedComponent@http://localhost:8081/index.bundle?platform=ios&dev=true&minify=false&modulesOnly=false&runModule=true&app=com.xxx.xxx:62612:24
AnimatedComponentWrapper
RCTView
View
Card@http://localhost:8081/index.bundle?platform=ios&dev=true&minify=false&modulesOnly=false&runModule=true&app=com.xxx.xxx:313751:36
CardContainer@http://localhost:8081/index.bundle?platform=ios&dev=true&minify=false&modulesOnly=false&runModule=true&app=com.xxx.xxx:313428:22
RNSScreen
AnimatedComponent@http://localhost:8081/index.bundle?platform=ios&dev=true&minify=false&modulesOnly=false&runModule=true&app=com.xxx.xxx:62612:24
AnimatedComponentWrapper
Screen@http://localhost:8081/index.bundle?platform=ios&dev=true&minify=false&modulesOnly=false&runModule=true&app=com.xxx.xxx:311981:36
MaybeScreen@http://localhost:8081/index.bundle?platform=ios&dev=true&minify=false&modulesOnly=false&runModule=true&app=com.xxx.xxx:312330:24
RNSScreenContainer
ScreenContainer@http://localhost:8081/index.bundle?platform=ios&dev=true&minify=false&modulesOnly=false&runModule=true&app=com.xxx.xxx:312072:36
MaybeScreenContainer@http://localhost:8081/index.bundle?platform=ios&dev=true&minify=false&modulesOnly=false&runModule=true&app=com.xxx.xxx:312299:23
CardStack@http://localhost:8081/index.bundle?platform=ios&dev=true&minify=false&modulesOnly=false&runModule=true&app=com.xxx.xxx:311455:36
KeyboardManager@http://localhost:8081/index.bundle?platform=ios&dev=true&minify=false&modulesOnly=false&runModule=true&app=com.xxx.xxx:316336:36
RNCSafeAreaProvider
SafeAreaProvider@http://localhost:8081/index.bundle?platform=ios&dev=true&minify=false&modulesOnly=false&runModule=true&app=com.xxx.xxx:308043:24
SafeAreaProviderCompat@http://localhost:8081/index.bundle?platform=ios&dev=true&minify=false&modulesOnly=false&runModule=true&app=com.xxx.xxx:316778:24
RCTView
View
StackView@http://localhost:8081/index.bundle?platform=ios&dev=true&minify=false&modulesOnly=false&runModule=true&app=com.xxx.xxx:307507:36
StackNavigator@http://localhost:8081/index.bundle?platform=ios&dev=true&minify=false&modulesOnly=false&runModule=true&app=com.xxx.xxx:307388:32
AuthenticationStack
EnsureSingleNavigator@http://localhost:8081/index.bundle?platform=ios&dev=true&minify=false&modulesOnly=false&runModule=true&app=com.xxx.xxx:137851:24
BaseNavigationContainer@http://localhost:8081/index.bundle?platform=ios&dev=true&minify=false&modulesOnly=false&runModule=true&app=com.xxx.xxx:137485:28
ThemeProvider@http://localhost:8081/index.bundle?platform=ios&dev=true&minify=false&modulesOnly=false&runModule=true&app=com.xxx.xxx:141746:21
NavigationContainer@http://localhost:8081/index.bundle?platform=ios&dev=true&minify=false&modulesOnly=false&runModule=true&app=com.xxx.xxx:141652:26
Provider@http://localhost:8081/index.bundle?platform=ios&dev=true&minify=false&modulesOnly=false&runModule=true&app=com.xxx.xxx:109664:26
Provider@http://localhost:8081/index.bundle?platform=ios&dev=true&minify=false&modulesOnly=false&runModule=true&app=com.xxx.xxx:302526:26
StripeProvider@http://localhost:8081/index.bundle?platform=ios&dev=true&minify=false&modulesOnly=false&runModule=true&app=com.xxx.xxx:299862:41
App@http://localhost:8081/index.bundle?platform=ios&dev=true&minify=false&modulesOnly=false&runModule=true&app=com.xxx.xxx:149198:20
Provider@http://localhost:8081/index.bundle?platform=ios&dev=true&minify=false&modulesOnly=false&runModule=true&app=com.xxx.xxx:144062:26
AuthProvider@http://localhost:8081/index.bundle?platform=ios&dev=true&minify=false&modulesOnly=false&runModule=true&app=com.xxx.xxx:142753:65
AppContainer@http://localhost:8081/index.bundle?platform=ios&dev=true&minify=false&modulesOnly=false&runModule=true&app=com.xxx.xxx:142655:68
Provider@http://localhost:8081/index.bundle?platform=ios&dev=true&minify=false&modulesOnly=false&runModule=true&app=com.xxx.xxx:109664:26
AppConfig
RCTView
View
RCTView
View
AppContainer@http://localhost:8081/index.bundle?platform=ios&dev=true&minify=false&modulesOnly=false&runModule=true&app=com.xxx.xxx:69675:36
```

**Comment by adrai** [comment]
is this really the first log statement of i18next? => `WARN  react-i18next:: i18n.languages were undefined or empty undefined`

There should be something like: `i18next: initialized` before that

**Comment by azuken** [comment]
@adrai Yes, I didn't see it sorry (I have other logs between). I have updated my previous comment.

**Comment by adrai** [comment]
And there is no other log statement before?

**Comment by adrai** [comment]
btw: please log also your translations.

i.e. console.log(enTranslation)

**Comment by azuken** [comment]
@adrai No, before init log I only have metro Bundle log :

` BUNDLE  ./index.js`

Here is an extract from console log for enTranslation import :

`LOG  {"booking": {"afternoon": "Afternoon", "beginning": "Beginning", "cancel": "Cancel", "cancelBooking": "Cancel booking", [...]}`

**Comment by adrai** [comment]
Out of curiosity: does it work if you use a custom languageDetector like this?

```js
const languageDetector = {
  type: 'languageDetector',
  async: true,
  detect: cb => cb('en'),
  init: () => {},
  cacheUserLanguage: () => {},
};

i18next
  .use(languageDetector)
  .use(initReactI18next)
  .init()
```

**Comment by azuken** [comment]
I've replaced my current language detector by your code, and I have no more warning by i18next.

I have now a log before initialization :

` LOG  i18next: languageChanged en`

My app is showing translations correctly, but do not select the right language (my real device is in french and is showing english).

But I see a Typescript error with this language dectector :

```
TS2345: Argument of type '{ type: string; async: boolean; detect: (cb: any) => any; init: () => void; cacheUserLanguage: () => void; }' is not assignable to parameter of type 'Module | ThirdPartyModule[] | Newable<ThirdPartyModule>[] | Newable<Module>'.   Type '{ type: string; async: boolean; detect: (cb: any) => any; init: () => void; cacheUserLanguage: () => void; }' is not assignable to type 'Module'.     Types of property 'type' are incompatible.       Type 'string' is not assignable to type '"languageDetector" | "3rdParty" | "backend" | "logger" | "postProcessor" | "i18nFormat"'.
```

**Comment by adrai** [comment]
Sorry, I'm not a react-native developer....

Maybe i18next-react-native-language-detector is outdated...
But looking at the code, it seems it directly depends on react-native-locale-detector: https://github.com/DylanVann/i18next-react-native-language-detector/blob/master/index.js

Maybe you can use react-native-localize: https://github.com/zoontek/react-native-localize#getlocales

**Comment by azuken** [comment]
Sorry for the delay,

I've tried to make my own language detector with localize, and it works perfectly ! No more warnings.

Here is the full code sample :

```typescript
import i18n, { LanguageDetectorModule } from 'i18next'
import * as RNLocalize from 'react-native-localize'
import enTranslation from './en/translation.json'
import frTranslation from './fr/translation.json'
import esTranslation from './es/translation.json'
import deTranslation from './de/translation.json'
import { initReactI18next } from 'react-i18next'

import phoneLocale from './language'
import moment from 'moment'
import 'moment/locale/fr'
import 'moment/locale/de'
import 'moment/locale/es'

moment.locale(phoneLocale)

const languageDetector: LanguageDetectorModule = {
  init: function () {
    return
  },
  type: 'languageDetector',
  detect: () => {
    const locales = RNLocalize.getLocales()
    if (locales && locales.length) return locales[0].languageCode
  },
  cacheUserLanguage: function () {
    return
  },
}

const resources = {
  en: { translation: enTranslation },
  fr: { translation: frTranslation },
  es: { translation: esTranslation },
  de: { translation: deTranslation },
}

if (!i18n.isInitialized)
  void i18n
    .use(initReactI18next)
    .use(languageDetector)
    .init({
      resources,
      fallbackLng: 'en',
      interpolation: {
        escapeValue: false,
      },
    })

export default i18n
```

**Comment by DoctorJohn** [comment]
@adrai I basically have the same issue. I'm using a custom language detector.

The issue occures when I let my language detectors `detect` method return `undefined`. According to the `LanguageDetectorModule` typescript interface returning `undefined` is okay. I assumed that i18next would default to the fallback language (`fallbackLng`) in this case. Instead the `react-i18next:: i18n.languages were undefined or empty undefined` warning appears.

While reading this issue I tried implementing an async language detector. I noticed that the callback passed to the `detect` method cannot be called with `undefined` as parameter. Now I'm wondering: Is a language detector allowed to return `undefined` from `detect` and will i18next default to `fallbackLng` in this case? Or is the language detector supposed to fallback by itself?

**Comment by adrai** [comment]
@DoctorJohn in theory it is ok, to return undefined for in the detect method.
The fallbackLng (if defined) should work.

**Comment by DoctorJohn** [comment]
@adrai looking at the code this appears to be true. But I still think something is wrong here.

I lost focus on the actual warning for a moment: `react-i18next:: i18n.languages were undefined or empty undefined `. So the question is why is i18n.languages undefined in this case.

I noticed `i18n.languages` is only set from the following two placed in the code:

- https://github.com/i18next/i18next/blob/5ccefb6e77bc4d2e13ba577a857d8cccbf756765/src/i18next.js#L268
- https://github.com/i18next/i18next/blob/5ccefb6e77bc4d2e13ba577a857d8cccbf756765/src/i18next.js#L288

The first one is within the `changeLanguage` method, the second one is within a `setLng` method. `changeLanguage` is unlikely to cause my problem, `setLng` however is. I noticed that `i18n.languages` is actually never set when the language detector returns `undefined`. I pasted the "suspicious" code below and added "steps" to show why:

```js
    // STEP 3: setLng gets called with `undefined``
    const setLng = lngs => {
      // STEP 4: lngs is `undefined` getBestMatchFromCodes returns null (see impl.)
      const l = typeof lngs === 'string' ? lngs : this.services.languageUtils.getBestMatchFromCodes(lngs);

      // STEP 5: `l` is null and thus `this.languages` will never be set
      if (l) {
        if (!this.language) {
          this.language = l;
          this.languages = this.services.languageUtils.toResolveHierarchy(l);
        }
        if (!this.translator.language) this.translator.changeLanguage(l);

        if (this.services.languageDetector) this.services.languageDetector.cacheUserLanguage(l);
      }

      this.loadResources(l, err => {
        done(err, l);
      });
    };

    if (!lng && this.services.languageDetector && !this.services.languageDetector.async) {
      // STEP 1: the language detector returns `undefined`
      // STEP 2: setLng is called with `undefined`
      setLng(this.services.languageDetector.detect());
    } else if (!lng && this.services.languageDetector && this.services.languageDetector.async) {
      this.services.languageDetector.detect(setLng);
    } else {
      setLng(lng);
    }
```

The code above is copied from here: https://github.com/i18next/i18next/blob/5ccefb6e77bc4d2e13ba577a857d8cccbf756765/src/i18next.js#L281

`i18n.languages` being undefined then causes useTranslations to fail.

**Comment by adrai** [comment]
Can you please provide a small reproducible example? It does not need to be react-native.

In general, if there is no language set, and no language detected, then there is something wrong somewhere.

**Comment by adrai** [comment]
@DoctorJohn Seems it works if the language detector returns an empty array only. Can you try with i18next v20.2.3?

**Comment by DoctorJohn** [comment]
@adrai I can confirm returning undefined in from a language detector works in v20.2.3, thanks.

Nitpick: As mentioned earlier, I noticed an inconsistency between the `LanguageDetectorModule` and `LanguageDetectorAsyncModule` typescript definitions. The sync `detect` method is allowed to return `string | undefined` but the async `detect` methods callback must resolve to a `string`. Now that falsy values are handled, I think the callback should be allowed to resolve to `string | undefined`.

https://github.com/i18next/i18next/blob/61d75cd83e84190d932f0b3bc3775b913fe28a59/index.d.ts#L794

https://github.com/i18next/i18next/blob/61d75cd83e84190d932f0b3bc3775b913fe28a59/index.d.ts#L809

**Comment by adrai** [comment]
In theory it's string or array...
//cc @pedrodurek

**Comment by adrai** [comment]
Types fixed with: v20.2.4
